### ISSUE TITLE
Director Module only works with configuration at /etc/icingaweb2

### ISSUE THREAD

**Issue by bahrpatrick** [Icinga/icingaweb2-module-director#2024]
## Expected Behavior

Director module has to check the environment variable in the web config.

## Current Behavior

The director background daemon has the error "DB configuration not valid"

Icinga Director requires configration at /etc/icingaweb2. 
Seems like the environment variable ICINGAWEB_CONFIGDIR set in the web configuration is not read by the director module.

## Possible Solution

Create a symlink at /etc/icingaweb2 to the config dir.

## Your Environment

* Director version (System - About): 1.7.2
* Icinga Web 2 version and modules (System - About): 2.7.3
* Icinga 2 version (`icinga2 --version`): 2.11.2
* Operating System and version: CentOS 7.7
* Webserver, PHP versions:

Hi,

I´ve just noticed a bug in the icinga director module when updating icingaweb2.
Please read above.

Thanks

**Comment by Thomas-Gelf** [comment]
Setting `ICINGAWEB_CONFIGDIR` should work just fine. What do you mean by

> [..] the environment variable ICINGAWEB_CONFIGDIR set in the web configuration [..]

If you want to set environment variables for the daemon, you need to provide it to the daemons environment. On CentOS, you can create the file `/etc/sysconfig/icingadirector` for this purpose

**Comment by bahrpatrick** [comment]
Hi Thomas,

thanks for your answer.
I mean the environment variable ICINGAWEB_CONFIGDIR is set in /etc/httpd/conf.d/icingaweb2.conf but shouldn´t the director module mind this variable?

Or do I need to set it seperate in the /etc/sysconfig/icingadirector file if my Icingaweb configdirectory is no installed to the default location /etc/icingaweb2.

Thanks for your support!

**Comment by arogic** [comment]
Hi @bahrpatrick ,

I also get the error "icingadirector[31099]: DB configuration is not valid" on my Ubuntu 18.04 Setup. Can you explain the "Possible Solution - Create a symlink at /etc/icingaweb2 to the config dir." Can you provide "Source" and "Destination" for the symlink?

Thanks in advance. Regards

**Comment by bahrpatrick** [comment]
Hi @arogic,

you need to link your icingaweb2 installation directory to /etc/icingaweb2.

For example: ln -s [YOUR ICINGAWEB2 CONFIG DIRECTORY] /etc/icingaweb2

The director database should be the correct database.

Hopefully this will fix your problem.

Regards

**Comment by arogic** [comment]
Hi @bahrpatrick ,

Thanks for your prompt response. I have updated my previous comment. My icingaweb2 config driectory is "/etc/icingaweb2" or  I don't understand properly. I am using apache and the icingaweb2.conf is in /etc/apache2/conf-available. Does it mean I have to create a symlink in my /etc/icingaweb2 driectory to the "/etc/apache2/conf-available/icingaweb2.conf"?

Thanks

**Comment by bahrpatrick** [comment]
Hi @arogic 

In my case the icingaweb2 config directory is based on another location because of this I linked my configuration directory to the default directory location.
This worked for me.

I think your problem is based elsewhere if you´re running default settings ....

Maybe you can post some more information about your configuration. 

Regards
Patrick

**Comment by arogic** [comment]
Hi @bahrpatrick ,

what information do you need? 
> icinga Web 2 Version: 2.7.3
> icinga2 - The Icinga 2 network monitoring daemon (version: r2.11.3-1)
> System information:
>  Platform: Ubuntu
>  Platform version: 18.04.4 LTS (Bionic Beaver)
>  Kernel: Linux
>  Kernel version: 4.15.0-96-generic
>  Architecture: x86_64
> General paths:
>  Config directory: /etc/icinga2
>  Data directory: /var/lib/icinga2
>  Log directory: /var/log/icinga2
>  Cache directory: /var/cache/icinga2
>  Spool directory: /var/spool/icinga2
>  Run directory: /run/icinga2
>
> Internal paths:
>  Package data directory: /usr/share/icinga2
>  State path: /var/lib/icinga2/icinga2.state
>  Modified attributes path: /var/lib/icinga2/modified-attributes.conf
>  Objects path: /var/cache/icinga2/icinga2.debug
>  Vars path: /var/cache/icinga2/icinga2.vars
>  PID path: /run/icinga2/icinga2.pid

As I mentioned before when I run:

`icingacli director daemon run --debug --trace`

the background daemon is running fine and it works. However it doesn't work when using systemd. I followed the icinga director documentation to setup the background daemon to use systemd.

Thanks in advance. Regards

**Comment by bahrpatrick** [comment]
Hi @arogic 

could you  maybe post your systemd service configuration?
And the error output while starting it with systemd.

Regards
Patrick

**Comment by arogic** [comment]
Hi @bahrpatrick ,

systemd service configuration:

> cat /etc/systemd/system/icinga-director.service
> [Unit]
> Description=Icinga Director - Monitoring Configuration
> Documentation=https://icinga.com/docs/director/latest/
> Wants=network.target
>
> [Service]
> EnvironmentFile=-/etc/default/icinga-director
> EnvironmentFile=-/etc/sysconfig/icinga-director
> ExecStart=/usr/bin/icingacli director daemon run
> ExecReload=/bin/kill -HUP ${MAINPID}
> User=icingadirector
> SyslogIdentifier=icingadirector
> Type=notify
>
> NotifyAccess=main
> WatchdogSec=10
> RestartSec=30
> Restart=always
>
> [Install]
> WantedBy=multi-user.target

systemd output when running "systemctl status icinga-director"

> user@host/etc/icingaweb2# systemctl status icinga-director
> ● icinga-director.service - Icinga Director - Monitoring Configuration
>   Loaded: loaded (/etc/systemd/system/icinga-director.service; enabled; vendor preset: enabled)
>   Active: active (running) since Mon 2020-04-20 10:11:44 UTC; 56min ago
>     Docs: https://icinga.com/docs/director/latest/
> Main PID: 31614 (icingacli)
>   Status: "running, db: no configuration"
>    Tasks: 1 (limit: 2361)
>   CGroup: /system.slice/icinga-director.service
>           └─31614 icinga::director: running, db: no configuration
>
> Apr 20 10:11:44 icinga2dev3 systemd[1]: Stopped Icinga Director - Monitoring Configuration.
> Apr 20 10:11:44 icinga2dev3 systemd[1]: Starting Icinga Director - Monitoring Configuration...
> Apr 20 10:11:44 icinga2dev3 icingadirector[31614]: Started by systemd, notifying watchdog every 5s via /run/sy
> Apr 20 10:11:44 icinga2dev3 systemd[1]: Started Icinga Director - Monitoring Configuration.
> Apr 20 10:11:44 icinga2dev3 icingadirector[31614]: DB configuration is not valid

Regards

**Comment by bahrpatrick** [comment]
Hi @arogic ,

from my point of view this is looking good.
Do you have the correct database as resource configured in the /etc/icingaweb2/modules/director/config.ini file?

in my case this is:

`[db]
resource = "icinga_director_db"
`

Best regards
Patrick

**Comment by arogic** [comment]
Hi @bahrpatrick ,

yes, the correct Database is configured - see below:

>user@host:/etc/icingaweb2/modules/director# cat config.ini
; /etc/icingaweb2/modules/director/config.ini
[db]
resource = "director_db"

I noticed that my director directory has some funny  permissions set - see below:

> user@host:/etc/icingaweb2/modules# ls -ld director/
d-ws-w--wT 2 www-data www-data 4096 Apr 20 13:23 director/

also take a look at my resources.ini - see below:

>root@icinga2dev3:/etc/icingaweb2# cat resources.ini
[icingaweb_db]
type = "db"
db = "mysql"
host = "host.example.com"
port = ""
dbname = "icingaweb2_db"
username = "icingaweb2_db_user"
password = "**************"
charset = "utf8"
use_ssl = "0"

[icinga_ido]
type = "db"
db = "mysql"
host = "host.example.com"
port = ""
dbname = "icinga2_db"
username = "icinga2_db_user"
password = "**************"
charset = "utf8"
use_ssl = "0"

[director_db]
type = "db"
db = "mysql"
host = "host.example.com"
port = ""
dbname = "director"
username = "director"
password = "**************"
charset = "utf8"
use_ssl = "0"

Thanks. Regards

**Comment by arogic** [comment]
Hi @bahrpatrick ,

I got it fixed. To resolve the issue I executed the following commands:

> usermod -a -G nagios,www-data icingadirector
chmod 0755 /etc/icingaweb2/modules/director
systemctl restart icinga2
systemctl start icinga-director

It seems that the issue was caused by the funny permissions I had  on the /etc/icingaweb2/modules/director directory. I also had to add the "icingadirector" user to groups "nagios " and "www-data" to get it working. After  restarted the "icinga2" service and then started the "icinga-director" everything started working. Thanks for you help. Regards
